First version of the certificate manager.

smarnach opened the pull request:
This code is based on the [old version in the load-balancer role](https://github.com/open-craft/ansible-load-balancer/blob/master/templates/manage_certs.py), but heavily modified.  I haven't included the old version in this repo since the diff wouldn't be useful anyway – basically all of the code has changed, or at least has different indentation levels now.

It has been split out into a separate repository for several reasons.  First, we will run a separate certificate managment server – with redundant load balancers it does not make sense to handle this on the load balancers anymore.  Second, I wanted to structure the code into multiple files to make it more readable.  And third, it has always been awkward to have Python code live in an Ansible repository and setting up the tests there.  For just a single file it kind of worked OK, but having it in a separate repository makes testing and developing it nicer in some respects.

### Changes compared to the old version

* Obviously, I added support to load configuration from Consul and store certificates back to Consul.
* I factored the parts talking to Consul out into interface classes, to it would be easy to support the old operation mode from the same code base.
* I created an class isolating all the parts interfacing with Certbot.  These parts were also adapted to the latest version of Consul, making use of the new features.  Certbot now supports listing and deleting certificates, so I could remove the code we had for these operations before.
* Another feature that didn't exist when the original version was written is deploy hooks, which allow us to get rid of the awkward "cert-watcher" job, with all its locking and subtleties.  The deploy hook will be called synchronously while running Certbot, so once the run is complete the new certificates will already be installed in Consul.  This requires a separate `deploy_cert.py` command, but with the new code structure this command can share some code with the certificate management script.
* The logic to decide whether to request new certificates has been changed.  We don't contact the load balancers to detect whether they have valid certificates anymore, but instead compare the desired domains with the certificates that are in Consul.  This may result in requests for new certificates in a few cases where we wouldn't have requested a new cert in the old version, but in general this should not be a problem.
* The logic to decide what certificates to drop does not check for wildcard certificates anymore.  The old version assumed that there may be manually added certificates in `/etc/haproxy/certs`, and was careful not to remove these.  The new version assumes it owns all certificates in Consul.  If we ever need to manually add certificates again, we should directly add them to the load balancers in `/etc/haproxy/certs`, possibly with Ansible, but the load balancer should not need to deal with them.

### Notes and concerns

* We need to configure ACLs in Consul, before deploying this.  It's currenlty OK, but I definitely don't want to give access to _all_ TLS certificates to all servers.
* We have to make sure to run the certificate manager both via cron and as a Consul watch.  The Consul watch is good to quickly react to configuration changes, while the (infrequent) cron job is required to retry certificate generation for failed attempts, e.g. when we are waiting for DNS to propagate.
* There are hardly any unit tests so far.  I will add first tests over the weekend, but much of the code base doesn't lend itself well to unit testing.
* The old integration tests were built into the load-balancer role.  It would be nice to have integration tests again, but I haven't thought about how to include them in this repo yet, and they won't be included in this PR.
* The `pylintrc` is essentially copied from Ocim, as requested by Xavier.
* I originally wrote `deploy_cert` as a [shell script](https://github.com/open-craft/cert-manager/blob/deploy-cert-shell/deploy_cert.sh).  That version was actually a lot shorter, but I prefer the Python version anyway, since it integrates much better with the rest of the code base, and because it's not shell.
* It would probably be good to add a simple `setup.py`, so the certificate manager can be installed system-wide via pip.

### Testing instructions

I've installed this to a stage server for testing.  Testing is still a bit awkward, since we don't have all the components in place yet.  Ocim already writes all instance information to Consul, but the ACME challenges are not yet forwarded to the new certificate manager, so we can't request certificates for actual instances, but only for DNS names pointing to the server itself, and only if we manually write this information to Consul in a non-standard location.

The test server is at `cert-manager-stage.net.opencraft.hosting`, and you should be able to log in as `ubuntu` with the usual VM identity (using the SSH config from [this branch](https://github.com/open-craft/ansible-secrets/tree/smarnach/cert-manager) should work as well].  I've also added the DNS name `cert-manager-stage-test.net.opencraft.hosting` pointing to the same IP address, so we have at least two domain names to play with.

Suggested testing steps:

0. Clean up potential leftovers from earlier tests:

       consul kv delete -recurse certs
       consul kv delete -recurse ocim-test
       sudo rm -rf /etc/letsencrypt

1. Add some domains to Consul, under the `ocim-test` prefix:

       consul kv put ocim-test/instances/1/domains '["cert-manager-stage.net.opencraft.hosting"]'
       consul kv put ocim-test/instances/2/domains '["cert-manager-stage-test.net.opencraft.hosting"]'

2. Run the certificate manager:

       cd cert-manager
       sudo pipenv run python manage_certs.py --contact-email ops@opencraft.com --letsencrypt-use-staging --log-level debug --webroot-path /var/www/certbot --consul-ocim-prefix ocim-test/instances

3. Verify the certificates

       consul kv get -recurse certs
       sudo pipenv run certbot certificates

4. Remove one of the domains from Consul, and run the cert manager again:

       consul kv delete -recurse ocim-test/instances/1
       sudo pipenv run python manage_certs.py --contact-email ops@opencraft.com --letsencrypt-use-staging --log-level debug --webroot-path /var/www/certbot --consul-ocim-prefix ocim-test/instances

5. Verify that the certificate has been removed from both Certbot and Consul (as in step 3).

You can also first start a root shell in the pipenv using `sudo pipenv shell`, which may save some typing.

Of course there are a lot more things you could test – these are some tests to verify the basics.

UmanShahzad commented on cert_manager/__init__.py:
```diff
@@ -0,0 +1,85 @@
+    def get_current_domains(self):
+        """Return the domain names of all certificates in the storage backend.
+
+        The return value is a dictionary mapping each certificate name to the list of all domain
```
Nit: It's technically mapping to a `set` rather than a `list`, so the available operations would be different.

UmanShahzad commented on cert_manager/backends.py:
```diff
@@ -0,0 +1,163 @@
+    def __init__(self, prefix):
+        """Initialize the backend with the given Consul prefix."""
+        self.prefix = prefix
+
```
Nit: extra newline.

UmanShahzad commented on cert_manager/certbot.py:
```diff
@@ -0,0 +1,100 @@
+"""The main CertificateManager class that interacts with Certbot."""
```
Nit: bad copy pasta :p

smarnach replied:
Not even that – I renamed both the class and the file and moved some code out, but did not update the docstring.

smarnach commented on Pipfile:
```diff
@@ -0,0 +1,17 @@
+[packages]
+certbot = "*"
+python-consul = "*"
+pyopenssl = "*"
```
I intentionally did not pin the packets, since this way we can easily update package versions later.  At least for Certbot, we should update to the latest version every now and then, in case they fix issues with the ACME protocol or have general security fixes.  I've at least specified minimum versions now, which will allow easy updates but at least provide some more explicit information in the Pipfile.

UmanShahzad replied:
Sounds good, thanks for the min-pin!

smarnach commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+.pytest_cache/
+__pycache__/
```
I don't think that's necessary anymore.  For Python 3.2 and up, all `.pyc` files go into the `__pycache__` directories, so excluding these should be enough.

UmanShahzad replied:
Ah OK!

smarnach commented on cert_manager/__init__.py:
```diff
@@ -0,0 +1,85 @@
+
+import logging
+
+from .utils import extract_x509_dns_names
```
I prefer explicit relative imports over absolute imports, since they are less verbose and less repetitive, and allow moving things around within a package without adjusting all imports.  I'm not aware of any downsides of explicit relative imports.

Python 2 has _implicit_ relative imports (without the leading dot), which made it rather ambiguous how certain reports were resolved.  In Python 2, I recommend always using `from __future__ import absolute_import`, which will disable implicit relative imports.  Python 3 doesn't have them anymore anyway.

UmanShahzad replied:
Ah I remember reading about this, but that explains it a lot better. Sounds good then.

smarnach commented on cert_manager/__init__.py:
```diff
@@ -0,0 +1,85 @@
+        """Remove a certificate from both Certbot and the backend storage."""
+        self.certbot.remove_cert(main_domain)
+        try:
+            self.cert_storage.remove_cert(main_domain)
```
In my opinion, the better approach is to raise exceptions and let the caller decide what to do with them, which is why I used this approach for the Consul backend.  The code in the Certbot client was partially copied from the old version, which caused this inconsitency.  I will change the Certbot client to match the Consul backend.

smarnach commented:
@UmanShahzad Thanks for the review!  I've addressed your comments, added further tests and set up CircleCI, so this should be complete now.  Could you please take another look?

UmanShahzad commented on cert_manager/backends.py:
```diff
@@ -0,0 +1,163 @@
+        The dictionary maps the main domain names to the PEM data of each certificate.
+        """
+        consul_client = consul.Consul()
+        dummy, values = consul_client.kv.get(self.prefix, recurse=True)
```
Sounds good.

smarnach commented:
@UmanShahzad Thanks!  I've made two minor changes (add documentation on running tests to README, don't bother installing pipenv on CircleCI since it is already installed), so feel free to take a look at the last two commits.  Merging now anyway, since the changes were rather trivial.
